File: csharp/uwapi/world.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Unnatural
{
    using MyForceStatistics = Interop.UwMyForceStatistics;
    using ForeignPolicy = Interop.UwForeignPolicyComponent;
    using PolicyEnum = Interop.UwForeignPolicyEnum;
    using UnitUpgrades = Interop.UwUnitUpgrades;
    using PathStateEnum = Interop.UwPathStateEnum;
    using OverviewFlags = Interop.UwOverviewFlags;

    public static class World
    {
        public static uint MyForceId()
        {
            return myForceId;
        }

        public static MyForceStatistics MyForceStatistics()
        {
            return myForceStatistics;
        }

        public static PathStateEnum UnitPathState(uint unitId)
        {
            return Interop.uwUnitPathState(unitId);
        }

        public static UnitUpgrades UnitUpgrades(uint unitId)
        {
            UnitUpgrades data = new UnitUpgrades();
            Interop.uwUnitUpgrades(unitId, ref data);
            return data;
        }

        public static bool TestShooting(uint shooterId, uint targetId)
        {
            return Interop.uwTestShootingEntities(shooterId, targetId);
        }

        public static bool TestConstructionPlacement(uint constructionProto, uint position, uint recipeProto = 0)
        {
            return Interop.uwTestConstructionPlacement(constructionProto, position, recipeProto);
        }

        public static uint FindConstructionPlacement(uint constructionProto, uint position, uint recipeProto = 0)
        {
            return Interop.uwFindConstructionPlacement(constructionProto, position, recipeProto);
        }

        public static IReadOnlyList<OverviewFlags> OverviewFlags()
        {
            return overview;
        }

        public static OverviewFlags OverviewFlags(uint position)
        {
            return overview[(int)position];
        }

        public static uint[] OverviewEntities(uint position)
        {
            Interop.UwIds ns = new Interop.UwIds();
            Interop.uwOverviewIds(position, ref ns);
            return InteropHelpers.Ids(ns);
        }

        public static void UnitPathfinding(Action<UnitPathfindingResult> callback, uint startingPosition, uint goalPosition, uint unitPrototype, bool allowNearbyPosition = false, uint maxIterations = 0)
        {
            Action fin = () =>
            {
                Interop.UwUnitPathfindingResult tmp = new Interop.UwUnitPathfindingResult();
                Interop.uwRetrieveUnitPathfinding(ref tmp);
                UnitPathfindingResult res = new UnitPathfindingResult();
                res.path = InteropHelpers.Ids(tmp.path);
                res.state = tmp.state;
                callback(res);
            };
            Interop.UwUnitPathfindingQuery q = new Interop.UwUnitPathfindingQuery();
            q.startingPosition = startingPosition;
            q.goalPosition = goalPosition;
            q.unitPrototype = unitPrototype;
            q.allowNearbyPosition = allowNearbyPosition;
            q.maxIterations = maxIterations;
            q.taskUserData = UwapiTasks.InsertTask(fin);
            Interop.uwStartUnitPathfinding(ref q);
        }

        public static IReadOnlyDictionary<uint, Entity> Entities()
        {
            return entities;
        }

        public static Entity Entity(uint id)
        {
            return entities[id];
        }

        public static PolicyEnum Policy(uint force)
        {
            PolicyEnum val;
            return policies.TryGetValue(force, out val) ? val : PolicyEnum.None;
        }

        static uint myForceId;
        static MyForceStatistics myForceStatistics = new MyForceStatistics();
        static readonly Dictionary<uint, Entity> entities = new Dictionary<uint, Entity>();
        static readonly Dictionary<uint, PolicyEnum> policies = new Dictionary<uint, PolicyEnum>();
        static OverviewFlags[] overview = new OverviewFlags[0];

        static uint[] AllIds()
        {
            Interop.UwIds ids = new Interop.UwIds();
            Interop.uwAllEntities(ref ids);
            return InteropHelpers.Ids(ids);
        }

        static uint[] ModifiedIds()
        {
            Interop.UwIds ids = new Interop.UwIds();
            Interop.uwModifiedEntities(ref ids);
            return InteropHelpers.Ids(ids);
        }

        static void UpdateRemoved()
        {
            var allIds = new HashSet<uint>(AllIds());
            var removed = new List<uint>();
            foreach (uint id in entities.Keys)
                if (!allIds.Contains(id))
                    removed.Add(id);
            foreach (uint id in removed)
            {
                entities[id].Destroyed = true;
                entities.Remove(id);
            }
        }

        static void UpdateFresh()
        {
            foreach (Entity e in entities.Values)
                e.Fresh = false;
        }

        static void UpdateModified()
        {
            foreach (uint id in ModifiedIds())
            {
                if (entities.ContainsKey(id))
                {
                    entities[id].FetchComponents();
                }
                else
                {
                    Entity o = new Entity(id);
                    o.FetchComponents();
                    entities[id] = o;
                }
            }
        }

        static void UpdatePolicies()
        {
            policies.Clear();
            foreach (Entity e in entities.Values)
            {
                if (!e.ForeignPolicy.HasValue)
                    continue;
                ForeignPolicy fp = e.ForeignPolicy.Value;
                if (fp.forces[0] == myForceId)
                    policies[fp.forces[1]] = fp.policy;
                if (fp.forces[1] == myForceId)
                    policies[fp.forces[0]] = fp.policy;
            }
        }

        static void UpdateOverview(bool stepping)
        {
            if (stepping)
            {
                Interop.UwOverviewExtract ex = new Interop.UwOverviewExtract();
                Interop.uwOverviewExtract(ref ex);
                if (overview.Length != ex.count)
                    overview = new OverviewFlags[ex.count];
                if (ex.count > 0)
                    Marshal.Copy(ex.flags, (int[])(object)overview, 0, (int)ex.count);
            }
            else
                overview = new OverviewFlags[0];
        }

        static void Updating(object sender, bool stepping)
        {
            {
                Interop.UwMyPlayer tmp = new Interop.UwMyPlayer();
                Interop.uwMyPlayer(ref tmp);
                myForceId = tmp.forceEntityId;
            }
            Interop.uwMyForceStatistics(ref myForceStatistics);
            UpdateRemoved();
            UpdateFresh();
            UpdateModified();
            UpdatePolicies();
            UpdateOverview(stepping);
        }

        static World()
        {
            Events.Updating += Updating;
        }
    }

    public struct UnitPathfindingResult
    {
        public uint[] path;
        public PathStateEnum state;
    }
}

File: python/uwapi/world.py
from .interop import *
from .events import uw_events
from .entity import Entity
from .entity_update_components import entity_update_components


def _make_empty_UwMyForceStatistics() -> UwMyForceStatistics:
    return UwMyForceStatistics(0, 0, 0, 0, 0, 0)


def _make_empty_UwMyPlayer() -> UwMyPlayer:
    return UwMyPlayer(0, 0, False, False)


class World:
    _instance = None
    _my_player = _make_empty_UwMyPlayer()
    _my_force_statistics = _make_empty_UwMyForceStatistics()
    _entities: dict[int, Entity] = {}
    _policies: dict[int, UwForeignPolicyEnum] = {}
    _overview: list[UwOverviewFlags] = []

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            uw_events.on_update(cls._instance._update)
        return cls._instance

    def my_player_id(self) -> int:
        return self._my_player.playerEntityId

    def my_force_id(self) -> int:
        return self._my_player.forceEntityId

    def is_primary_controller(self) -> bool:
        return self._my_player.primaryController

    def is_admin(self) -> bool:
        return self._my_player.admin

    def my_force_statistics(self) -> UwMyForceStatistics:
        return self._my_force_statistics

    def unit_path_state(self, unit_id: int) -> UwPathStateEnum:
        return uw_interop.uwUnitPathState(unit_id)

    def unit_upgrades(self, unit_id: int) -> UwUnitUpgrades:
        return uw_interop.uwUnitUpgrades(unit_id)

    def test_shooting(self, shooter_id: int, target_id: int) -> bool:
        return uw_interop.uwTestShootingEntities(shooter_id, target_id)

    def test_construction_placement(
        self, construction_proto: int, position: int, recipe_proto: int = 0
    ) -> bool:
        return uw_interop.uwTestConstructionPlacement(
            construction_proto, position, recipe_proto
        )

    def find_construction_placement(
        self, construction_proto: int, position: int, recipe_proto: int = 0
    ) -> int:
        return uw_interop.uwFindConstructionPlacement(
            construction_proto, position, recipe_proto
        )

    def overview_flags_all(self) -> list[UwOverviewFlags]:
        return self._overview

    def overview_flags(self, position: int) -> UwOverviewFlags:
        return self._overview[position]

    def overview_entities(self, position: int) -> list[int]:
        return uw_interop.uwOverviewIds(position).ids

    def entities(self) -> dict[int, Entity]:
        return self._entities

    def entity(self, entity_id: int) -> Entity:
        return self._entities[entity_id]

    def policy(self, force_id: int) -> UwForeignPolicyEnum:
        return self._policies.get(force_id, UwForeignPolicyEnum.Nothing)

    def _all_ids(self) -> list[int]:
        return uw_interop.uwAllEntities().ids

    def _modified_ids(self) -> list[int]:
        return uw_interop.uwModifiedEntities().ids

    def _update_removed(self) -> None:
        all_ids = set(self._all_ids())
        removed = [eid for eid in self._entities if eid not in all_ids]
        for eid in removed:
            self._entities[eid].destroyed = True
            self._entities.pop(eid, None)

    def _update_fresh(self) -> None:
        for e in self._entities.values():
            e.fresh = False

    def _update_modified(self) -> None:
        for eid in self._modified_ids():
            if eid in self._entities:
                entity_update_components(self._entities[eid])
            else:
                e = Entity(eid)
                entity_update_components(e)
                self._entities[eid] = e

    def _update_policies(self) -> None:
        self._policies.clear()
        for e in self._entities.values():
            if not e.ForeignPolicy:
                continue
            fp = e.ForeignPolicy
            if fp.forces[0] == self._my_player.forceEntityId:
                self._policies[fp.forces[1]] = fp.policy
            if fp.forces[1] == self._my_player.forceEntityId:
                self._policies[fp.forces[0]] = fp.policy

    def _update_overview(self, stepping: bool) -> None:
        if stepping:
            self._overview = uw_interop.uwOverviewExtract().flags
        else:
            self._overview = []

    def _update(self, stepping: bool) -> None:
        tmp = uw_interop.uwMyPlayer()
        self._my_player = tmp[1] if tmp[0] else _make_empty_UwMyPlayer()
        self._my_force_statistics = uw_interop.uwMyForceStatistics()
        self._update_removed()
        self._update_fresh()
        self._update_modified()
        self._update_policies()
        self._update_overview(stepping)


uw_world = World()
